sound/soc/codecs/rt715-sdca.c

```diff
@@ -0,0 +1,938 @@
+	}
+
+	return ret;
+}
+
+static int rt715_index_read(struct rt715_sdca_priv *rt715,
+		unsigned int nid, unsigned int reg, unsigned int *value)
+{
+	struct regmap *regmap = rt715->regmap;
+	unsigned int addr;
+	int ret;
+
+	addr = (nid << 20) | reg;
+
+	ret = regmap_read(regmap, addr, value);
+	if (ret < 0)
+		pr_err("Failed to get private value: %06x => %04x ret=%d\n",
+			addr, *value, ret);
+
+	return ret;
+}
+
+static int rt715_index_update_bits(struct rt715_sdca_priv *rt715,
+	unsigned int nid, unsigned int reg, unsigned int mask, unsigned int val)
+{
+	unsigned int tmp;
+	int ret;
+
+	ret = rt715_index_read(rt715, nid, reg, &tmp);
+	if (ret < 0)
+		return ret;
+
+	set_mask_bits(&tmp, mask, val);
+
+	return rt715_index_write(rt715, nid, reg, tmp);
+}
+
+/* SDCA Volume/Boost control */
+static int rt715_set_amp_gain_put_sdca(struct snd_kcontrol *kcontrol,
+		struct snd_ctl_elem_value *ucontrol)
+{
+	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
+	struct soc_mixer_control *mc =
+		(struct soc_mixer_control *)kcontrol->private_value;
+	struct rt715_sdca_priv *rt715 = snd_soc_component_get_drvdata(component);
+	unsigned int val_l, val_r, gain_l_val, gain_r_val;
+	unsigned int i;
+	int ret_wl, ret_wr, ret_rl, ret_rr;
+
+	/* control value to 2s complement */
+	/* L channel */
+	gain_l_val = ucontrol->value.integer.value[0];
+	if (gain_l_val > mc->max)
+		gain_l_val = mc->max;
+	val_l = gain_l_val;
+
+	if (mc->shift == 8)
+		gain_l_val = (gain_l_val * 10) << mc->shift;
+	else {
+		gain_l_val = abs(gain_l_val - mc->shift) * RT715_SDCA_DB_STEP;
+		gain_l_val <<= 8;
+		gain_l_val /= 1000;
+		if (val_l <= mc->shift) {
+			gain_l_val = ~gain_l_val;
+			gain_l_val += 1;
+		}
+		gain_l_val &= 0xffff;
+	}
+
+	/* R channel */
+	gain_r_val = ucontrol->value.integer.value[1];
+	if (gain_r_val > mc->max)
+		gain_r_val = mc->max;
+	val_r = gain_r_val;
+
+	if (mc->shift == 8)
+		gain_r_val = (gain_r_val*10) << mc->shift;
+	else {
+		gain_r_val = abs(gain_r_val - mc->shift) * RT715_SDCA_DB_STEP;
+		gain_r_val <<= 8;
+		gain_r_val /= 1000;
+		if (val_r <= mc->shift) {
+			gain_r_val = ~gain_r_val;
+			gain_r_val += 1;
+		}
+		gain_r_val &= 0xffff;
+	}
+
+	for (i = 0; i < 3; i++) { /* retry 3 times at most */
+		/* Lch*/
+		ret_wl = regmap_write(rt715->regmap,
+				mc->reg, gain_l_val);
+		if (ret_wl != 0)
+			pr_err("Failed to write 0x%x=0x%x\n", mc->reg, gain_l_val);
+		/* Rch */
+		ret_wr = regmap_write(rt715->regmap,
+				mc->rreg, gain_r_val);
+		if (ret_wr != 0)
+			pr_err("Failed to write 0x%x=0x%x\n", mc->rreg, gain_r_val);
+
+		ret_rl = regmap_read(rt715->regmap, mc->reg, &val_l);
+		if (ret_rl < 0)
+			pr_err("Failed to read 0x%x, ret_rl=%d\n", mc->reg, ret_rl);
+		ret_rr = regmap_read(rt715->regmap, mc->rreg, &val_r);
+		if (ret_rr < 0)
+			pr_err("Failed to read 0x%x, ret_rr=%d\n", mc->rreg, ret_rr);
+
+		if (val_r == gain_r_val && val_l == gain_l_val)
+			break;
+	}
+
+	return (ret_wl || ret_wr || ret_rl || ret_rr);
```

Comment:
this would return 1 if any of those return codes were != 0, whereas you'd normally want a negative error code. Please, just use one `ret` variable as described above.